### PR TITLE
DuckPlayer: Overlays

### PR DESCRIPTION
**Pull request by shakyShane**
@jonathanKingston This pr is ready now, some important notes:

1) - this is a large PR, but is only included in windows, thanks to the recent change we made to bundling
2) - I've been extra careful to ensure this doesn't affect macOS + existing Duck Player
3) - I moved the messaging setup inside DuckPlayerFeature - mostly because it's the only consumer right now, but also because it bloats all bundles because of the extensive docs
  - let's fast-follow with comment-stripping to address this.
2) - I used the remote-config technique we discussed, but also if the config is absent I check the origins manually for launch. We can remove that aspect once it's merged into privacy config and live etc.

```json
{
  "unprotectedTemporary": [],
  "features": {
    "duckPlayer": {
      "state": "enabled",
      "exceptions": [],
      "settings": {
        "overlays": [
          {
            "domain": "www.youtube.com",
            "state": "enabled"
          }
        ],
        "proxy": []
      }
    }
  }
}

```

### CODE REVIEW

**Comment by shakyShane** [comment]
Current dependencies on/for this PR:
* main
  * **PR #412** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/412" target="_blank">Graphite</a> 
    * **PR #414** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/414" target="_blank">Graphite</a> 
      * **PR #417** <a href="https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/417" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/content-scope-scripts/417?utm_source=stack-comment).

File: package.json [hunk 1]
```diff
@@ -35,7 +35,12 @@
     "test": "npm run lint && npm run lint-no-output-globals && npm run test-unit && npm run test-int",
     "posttest": "npm run test --workspaces --if-present",
     "posttest-int": "npm run test-int --workspaces --if-present",
-    "posttest-int-x": "npm run test-int-x --workspaces --if-present"
+    "posttest-int-x": "npm run test-int-x --workspaces --if-present",
+    "serve": "http-server -c-1 --port 3220 integration-test/test-pages",
+    "playwright": "playwright test",
+    "playwright.headed": "playwright test --headed",
```
Comment (jonathanKingston) [inline package.json:41]: Can we keep with dashes or change the others?

File: integration-test/test-pages/duckplayer/config/overlays.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
+{
+  "unprotectedTemporary": [],
+  "features": {
+    "duckPlayer": {
+      "state": "enabled",
+      "exceptions": [],
+      "settings": {
+        "overlays": [
+          {
+            "domain": "www.youtube.com",
```
Comment (jonathanKingston) [inline integration-test/test-pages/duckplayer/config/overlays.json:10]: Please change this format

File: src/features/duck-player.js [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+ * @typedef UserValues - A way to communicate user settings
+ * @property {{enabled: {}} | {alwaysAsk:{}} | {disabled:{}}} privatePlayerMode - one of 3 values
+ * @property {boolean} overlayInteracted - always a boolean
+ */
+
+/**
+ * @param {Environment} environment - methods to read environment-sensitive things like the current URL etc
+ * @param {DuckPlayerOverlayMessages} comms - methods to communicate with a native backend
+ */
+async function initOverlays (environment, comms) {
```
Comment (jonathanKingston) [inline src/features/duck-player.js:33]: Can we throw this into it's own file under `features/duckplayer/....` ?

File: integration-test/test-pages/duckplayer/pages/overlays.html [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
...YAhQCpAMMApACbAKQAwwDsANMA0wDsASoBGwEqAYUBhQILEQAUABQAFAAUABUAFAAXABkAGQAXAB8AIgAeACIAHwAuACsAJwAnACsALgBGADIANgAyADYAMgBGAGoAQgBOAEIAQgBOAEIAagBeAHIAXQBWAF0AcgBeAKkAhQB2AHYAhQCpAMMApACbAKQAwwDsANMA0wDsASoBGwEqAYUBhQIL/8IAEQgAhQDIAwEiAAIRAQMRAf/EADEAAQEBAQEBAQAAAAAAAAAAAAABAgMEBQYBAQEBAAMAAAAAAAAAAAAAAAABAwIEBf/aAAwDAQACEAMQAAAA/LgFJbDXXh1XUspjcCjDQwpJNZJVjLYwCAqCgtlNxbY0Ioy1DLeSY3pMWyAMIAAKlFlNb56NXJdM0us96lpJQvDv44kQyAABZQC2DVzSp3N6Dt6fFcfRZs16M8Hq8t4alyQAAAAAFTZ09Npm0RRKh48QWIAAAAAAX3eX0nSzUSilE8/p+echYAAAAAAAA3gdunlHu6fNH0fnAAAAAAAAAAAAAAAAAAAAAAAAAAAAAB//xAAqEAABAwIFBAIBBQAAAAAAAAABAAIRAyEQEjFRYQQgMEFAQhMyYHKRsf/aAAgBAQABPwDuBVkQAiAoCjhQFAUBEBQFAUBQoUKFHcI9lFBNM3ML9RmwWpWpsEUeZlWj2vS4JQUI6C63OyGoVibYnsCMbyo5V4UgWCBtoFPCBwJE2lWIEEyuCpw/0qcIEXaUBbE+GAsoWULKsvKDSEQT6CiBohSeRMtHBRY7ZZXCbBGhWAvSciH7Py8hT5BhPawTf+uyVmdBEmCIIlVcrWwBrgfgAZjHr331jL42wPnubBNblEYNpzFrldR0YoED8rH/AMXSQnDKcCYBK1OB89NkCTqccztyg9yJJwrmGgboYHySpVJmYydPBWdLzxZDzsYXlAACB3uOVpO2EqfKASYCYzIPB1DoAbv8Ck9rDJCFWmfsgQdCO+s7NUPww940cUK9QewUOpPtqHUU/YIQq0z9wn1WNaYcJi37r//EABwRAAIABwAAAAAAAAAAAAAAAAERAAIDISJQYP/aAAgBAgEBPwDXmWpfKA0GeF//xAAeEQACAQMFAAAAAAAAAAAAAAABAhEDBCEAEjFQYP/aAAgBAwEBPwDr1uLOEBocDOAZ1UZWdii7VnA8L//Z';
```
Comment (jonathanKingston) [inline integration-test/test-pages/duckplayer/pages/overlays.html:87]: Remove?

File: packages/messaging/lib/windows.js [hunk 5]
```diff
@@ -32,6 +32,14 @@ export class WindowsMessagingTransport {
     constructor (config, messagingContext) {
         this.messagingContext = messagingContext
         this.config = config
+        this.globals = {
+            window,
+            JSONparse: window.JSON.parse,
```
Comment (jonathanKingston) [inline packages/messaging/lib/windows.js:37]: We should solve this in the abstract soon: https://app.asana.com/0/0/1204392417715156/f

File: src/features/duckplayer/icon-overlay.js [hunk 6]
```diff
@@ -0,0 +1,273 @@
+import { addTrustedEventListener, appendElement, VideoParams } from './util'
+import dax from './assets/dax.svg'
+import { i18n } from './text.js'
+import { OpenInDuckPlayerMsg } from './overlay-messages.js'
+
+export const IconOverlay = {
```
Comment (jonathanKingston) [inline src/features/duckplayer/icon-overlay.js:6]: Let's fast follow with making these as classes.

File: src/features/duckplayer/overlay-messages.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    getUserValues () {
+        return this.messaging.request(constants.MSG_NAME_READ_VALUES, {})
+    }
+
+    /**
+     * @param {Pixel} pixel
+     */
+    sendPixel (pixel) {
+        this.messaging.notify(constants.MSG_NAME_PIXEL, {
```
Comment (jonathanKingston) [inline src/features/duckplayer/overlay-messages.js:44]: Remember that once we consider rolling this for extensions we need to ensure that this is disabled for Firefox.

We should probably fast follow to ensure that firefox isn't enable-able even if this overlays file is bundled (which I know it's not right now)

File: src/features/duckplayer/components/ddg-video-overlay.js [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        shadow.appendChild(style)
+    }
+
+    /**
+     * @returns {HTMLDivElement}
+     */
+    createOverlay () {
+        const overlayElement = document.createElement('div')
+        overlayElement.classList.add('ddg-video-player-overlay')
+        overlayElement.innerHTML = `
```
Comment (jonathanKingston) [inline src/features/duckplayer/components/ddg-video-overlay.js:57]: Let's fast follow with that super lightweight HTML escaped lib that we talked about... please ping me to find it once ready.

File: src/features/duck-player.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+/**
+ * @internal
+ */
+export default class DuckPlayerFeature extends ContentFeature {
+    /**
+     * Lazily create a messaging instance for the given Platform + feature combo
+     * @return {Messaging}
+     */
+    get messaging () {
```
Comment (jonathanKingston) [inline src/features/duck-player.js:28]: As agreed we're going to put this in the `ContentFeature` prototype but it's not worth the bundle size changing for other builds yet.

Please fast follow here too, to clean this up.

File: src/features/duckplayer/components/ddg-video-overlay.js (hunk 8 above)
Comment (jonathanKingston) [reply src/features/duckplayer/components/ddg-video-overlay.js:57]: https://github.com/mozilla-extensions/secure-proxy/blob/db4d1b0e2bfe0abae416bf04241916f9e4768fd2/src/commons/template.js#L62 << this code

**Review by jonathanKingston — APPROVED** [review]
LGTM thanks for all the changes. We're in agreement that there's lot's of fast follows and issues to raise.


Thanks for making this happen!

**Pull request merged**